### ISSUE TITLE
Is the training of 2048 resolution only supported in HyperAttn and must rely on flash_attn at the same time?

### ISSUE THREAD

**Issue by zhenbuxianggaimingzi** [CastleChen339/DehazeXL#2]
Is the training of 2048 resolution only supported in HyperAttn and must rely on flash_attn at the same time? 
I try to replace the bottleneck of attn block, with scaled_dot_product_attention method from torch>=2.2.x instead of flash_attn, will OOM on 80 g A100~

![Image](https://github.com/user-attachments/assets/6efaaa8e-cd04-4f24-9ee4-4118b9cad835)

**Comment by CastleChen339** [comment]
是的，HyperAttn 和 FlashAttn 在显存优化方面确实起到了关键作用。我们在 A100 上测试了 2048×2048 分辨率的训练场景，即使使用HyperAttn，batch size 最大也只能到 2，显存占用大约为 67GB。如果你尝试用 PyTorch >=2.2.x 中的 `scaled_dot_product_attention` 替代HyperAttn，确实有可能出现 OOM（显存溢出）的情况，尤其是在更高分辨率和大batch 情况下。

如果你希望进一步优化显存使用，可以尝试一些最新的轻量稀疏注意力机制，没准能取得更好的效果。欢迎交流你的实验结果！


Yes, both HyperAttn and FlashAttn play a significant role in memory optimization. Based on our tests, when training with 2048×2048 resolution images on an A100 GPU, the maximum feasible batch size is 2, with around 67GB of memory usage—even with these optimizations in place. If you replace HyperAttn with the `scaled_dot_product_attention` from PyTorch >=2.2.x, it's quite possible to run into OOM issues under such high-resolution settings.

That said, you might want to explore some sparse attention mechanisms that are specifically designed to reduce memory usage. Some of them might even outperform FlashAttn/HyperAttn in certain high-resolution tasks. Would love to hear how your experiments go!

**Comment by fengyanzi** [comment]
You can try modifying the exact_attention function in the file DehazeXL/models/DehazeXL/context_encoders/hyper_attn/attention/utils.py as follows:
`def exact_attention(q, k, v, softmax_scale, causal=False, bias=None):
    qk = q @ k.transpose(-1, -2) * softmax_scale
    out = qk.softmax(dim=-1) @ v
    lse = torch.logsumexp(qk, dim=-1, keepdim=True)
    return out, lse`
This modification retains the functionality of hyperattention while removing the dependency on flash attention.

**Comment by fengyanzi** [comment]
@zhenbuxianggaimingzi 
如果解决了问题，欢迎及时反馈，由于部分代码直接采用HyperAttn，存在冗余，我会尝试提交PR修复这一问题
If the issue is resolved, please provide timely feedback. Since part of the code directly adopts HyperAttn, which leads to redundancy, I will attempt to fix this problem.

**Comment by zhenbuxianggaimingzi** [comment]
> 是的，HyperAttn 和 FlashAttn 在显存优化方面确实起到了关键作用。我们在 A100 上测试了 2048×2048 分辨率的训练场景，即使使用HyperAttn，batch size 最大也只能到 2，显存占用大约为 67GB。如果你尝试用 PyTorch >=2.2.x 中的 `scaled_dot_product_attention` 替代HyperAttn，确实有可能出现 OOM（显存溢出）的情况，尤其是在更高分辨率和大batch 情况下。
> 
> 如果你希望进一步优化显存使用，可以尝试一些最新的轻量稀疏注意力机制，没准能取得更好的效果。欢迎交流你的实验结果！
> 
> Yes, both HyperAttn and FlashAttn play a significant role in memory optimization. Based on our tests, when training with 2048×2048 resolution images on an A100 GPU, the maximum feasible batch size is 2, with around 67GB of memory usage—even with these optimizations in place. If you replace HyperAttn with the `scaled_dot_product_attention` from PyTorch >=2.2.x, it's quite possible to run into OOM issues under such high-resolution settings.
> 
> That said, you might want to explore some sparse attention mechanisms that are specifically designed to reduce memory usage. Some of them might even outperform FlashAttn/HyperAttn in certain high-resolution tasks. Would love to hear how your experiments go!

了解了，非常感谢~
还有个小问题，因为文中的消融实验没有针对分辨率的，如果将训练分辨率缩小到1024*1024，您有做过对比吗？这个性能相较于2048会下降多少呢？

**Comment by zhenbuxianggaimingzi** [comment]
> [@zhenbuxianggaimingzi](https://github.com/zhenbuxianggaimingzi) 如果解决了问题，欢迎及时反馈，由于部分代码直接采用HyperAttn，存在冗余，我会尝试提交PR修复这一问题 If the issue is resolved, please provide timely feedback. Since part of the code directly adopts HyperAttn, which leads to redundancy, I will attempt to fix this problem.

非常感谢~ 我正准备这样做改动，十分感谢~

**Comment by zhenbuxianggaimingzi** [comment]
明白了，非常感谢您的解答~
